### PR TITLE
Replace old REST clients with generated one

### PR DESCRIPTION
**Pull request by Y-Sindo**

### CODE REVIEW

File: test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs [hunk 1]
```diff
@@ -71,20 +76,60 @@ internal void GenerateClientEndpointTest(string appName, string expectedClientEn
...
+            }, null, restClient);
 
             using (var loggerFactory = useLoggerFacory ? (ILoggerFactory)new LoggerFactory() : NullLoggerFactory.Instance)
             {
                 var hubContext = await serviceManager.CreateHubContextAsync(HubName, loggerFactory);
             }
         }
 
+        [Fact]
+        internal async void IsServiceHealthy_ReturnTrue_Test()
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs:95]: Task

File: test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs [hunk 2]
```diff
@@ -60,7 +64,8 @@ internal void GenerateClientAccessTokenTest(string userId, Claim[] claims, strin
         [MemberData(nameof(TestGenerateClientEndpointData))]
         internal void GenerateClientEndpointTest(string appName, string expectedClientEndpoint)
         {
-            var manager = new ServiceManager(new ServiceManagerOptions() { ConnectionString = _testConnectionString, ApplicationName = appName }, null);
+            using var restClient = Mock.Of<ISignalRServiceRestClient>(client => client.BaseUri == new Uri("http://localhost"));
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs:67]: Endpoint?

File: test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs [hunk 3]
```diff
@@ -71,20 +76,60 @@ internal void GenerateClientEndpointTest(string appName, string expectedClientEn
...
+        [InlineData(HttpStatusCode.NotFound, typeof(AzureSignalRInaccessibleEndpointException))]
+        [InlineData(HttpStatusCode.Ambiguous, typeof(AzureSignalRRuntimeException))]
+        internal async void IsServiceHealthy_Throw_Test(HttpStatusCode statusCode, Type expectedException)
+        {
+            using ISignalRServiceRestClient restClient = new TestRestClient(statusCode);
+            var serviceManager = new ServiceManager(new ServiceManagerOptions(), null, restClient);
+
+            Task<bool> func() => serviceManager.IsServiceHealthy(default);
+
+            var exception = await Assert.ThrowsAnyAsync<AzureSignalRException>(func);
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs:129]: () => serviceManager.IsServiceHealthy(default)

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs [hunk 4]
```diff
@@ -40,7 +40,11 @@ public ServiceManagerBuilder WithCallingAssembly()
         public IServiceManager Build()
         {
             _options.ValidateOptions();
-            return new ServiceManager(_options, ProductInfo.GetProductInfo(_assembly));
+
+            string productInfo = ProductInfo.GetProductInfo(_assembly);
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:44]: Var

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs [hunk 5]
```diff
@@ -40,7 +40,11 @@ public ServiceManagerBuilder WithCallingAssembly()
...
+            ISignalRServiceRestClient restClient = new RestClientBuilder(_options.ConnectionString).WithUserAgent(productInfo).Build();
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:45]: Var

File: src/Microsoft.Azure.SignalR.Management/ServiceManager.cs [hunk 6]
```diff
@@ -1,22 +1,22 @@
-﻿// Copyright (c) Microsoft. All rights reserved.
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:1]: Revert

File: src/Microsoft.Azure.SignalR.Management/ServiceManager.cs [hunk 7]
```diff
@@ -160,27 +160,20 @@ public string GenerateClientAccessToken(string hubName, string userId = null, IL
...
-                },
-                cancellationToken: cancellationToken);
-            return isHealthy;
+            try
+            {
+                var healthApi = _restClient.HealthApi;
+                using HttpOperationResponse response = await healthApi.GetHealthStatusWithHttpMessagesAsync(cancellationToken: cancellationToken);
+                return true;
+            }
+            catch (HttpOperationException e) when ((int)e.Response.StatusCode >= 500 && (int)e.Response.StatusCode < 600)
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:169]: What about 505

File: src/Microsoft.Azure.SignalR.Management/ServiceManager.cs [hunk 8]
```diff
@@ -160,27 +160,20 @@ public string GenerateClientAccessToken(string hubName, string userId = null, IL
...
-                        default:
-                            return false;
-                    }
-                },
-                cancellationToken: cancellationToken);
-            return isHealthy;
+            try
+            {
+                var healthApi = _restClient.HealthApi;
+                using HttpOperationResponse response = await healthApi.GetHealthStatusWithHttpMessagesAsync(cancellationToken: cancellationToken);
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:166]: Var

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        public Exception WrapException(Exception e)
+        {
+            switch (e)
+            {
+                case HttpOperationException operationException:
+                    var response = operationException.Response;
+                    var request = operationException.Request;
+                    var detail = response.Content;
+
+                    var innerException = new HttpRequestException(
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs:29]: also put e into HttpRequestException as inner exception

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net;
+using System.Net.Http;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.SignalR.Common.RestClients
+{
+    public class HttpExceptionWrapper
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs:11]: make it a static exception extension class?

Comment (wanlwanl) [reply src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs:11]: usage: 
wrapper.WrapException(Exception e) -->
e.ConvertToAzureSignalRException()

@vicancy waht do you think?

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+                    return response.StatusCode switch
+                    {
+                        HttpStatusCode.BadRequest => new AzureSignalRInvalidArgumentException(request.RequestUri.ToString(), innerException, detail),
+                        HttpStatusCode.Unauthorized => new AzureSignalRUnauthorizedException(request.RequestUri.ToString(), innerException),
+                        HttpStatusCode.NotFound => new AzureSignalRInaccessibleEndpointException(request.RequestUri.ToString(), innerException),
+                        _ => new AzureSignalRRuntimeException(request.RequestUri.ToString(), innerException),
+                    };
+                case HttpRequestException requestException:
+                    return new AzureSignalRInaccessibleEndpointException(_baseUri.ToString(), requestException);
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs:40]: really necessary? can it be extracted from requestUri?

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs:40]: Request URI is invisible from HttpRequestException.

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs (hunk 10 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapper.cs]: I think this advice is good. Already done.

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs [hunk 12]
```diff
@@ -40,7 +40,11 @@ public ServiceManagerBuilder WithCallingAssembly()
         public IServiceManager Build()
         {
             _options.ValidateOptions();
-            return new ServiceManager(_options, ProductInfo.GetProductInfo(_assembly));
+
+            var productInfo = ProductInfo.GetProductInfo(_assembly);
+            ISignalRServiceRestClient restClient = new RestClientBuilder(_options.ConnectionString).WithUserAgent(productInfo).Build();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:45]: sounds that the `RestClientBuilder` can be a local field of `ServiceManagerBuilder`?

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    internal static class HttpExceptionWrapperExtension
+    {
+        internal static Exception WrapAsAzureSingalRException(this Exception e, Uri baseUri)
+        {
+            switch (e)
+            {
+                case HttpOperationException operationException:
+                    var response = operationException.Response;
+                    var request = operationException.Request;
+                    var detail = response.Content;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:20]: what would be inside `detail`? what if the server returns for example: `StatusCode(status, new { detail = "This is unhealthy" }`

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:20]: This is consistent with old REST client and it tries to give more information to users. But if server doesn't give more information other than status code, I don't think we can do more things.

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net;
+using System.Net.Http;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.SignalR.Common.RestClients
+{
+    internal static class HttpExceptionWrapperExtension
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:11]: Should be `ExceptionExtensions`

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            switch (e)
+            {
+                case HttpOperationException operationException:
+                    var response = operationException.Response;
+                    var request = operationException.Request;
+                    var detail = response.Content;
+
+                    var innerException = new HttpRequestException(
+                        $"Response status code does not indicate success: {(int)response.StatusCode} ({response.ReasonPhrase})", operationException); ;
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:23]: `; ;` --> `;`

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+                        HttpStatusCode.Unauthorized => new AzureSignalRUnauthorizedException(request.RequestUri.ToString(), innerException),
+                        HttpStatusCode.NotFound => new AzureSignalRInaccessibleEndpointException(request.RequestUri.ToString(), innerException),
+                        _ => new AzureSignalRRuntimeException(baseUri.ToString(), innerException),
+                    };
+                case HttpRequestException requestException:
+                    return new AzureSignalRInaccessibleEndpointException(baseUri.ToString(), requestException);
+                default:
+                    return e;
+            }
+
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:37]: extra blank line

File: test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs [hunk 17]
```diff
@@ -95,4 +134,4 @@ private static string GetExpectedClientEndpoint(string appName = null)
             return $"{Endpoint}/client/?hub={appName.ToLower()}_{HubName.ToLower()}";
         }
     }
-}
+}
```
Comment (wanlwanl) [inline test/Microsoft.Azure.SignalR.Management.Tests/ServiceManagerFacts.cs:137]: revert

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+using System;
+using System.Net;
+using System.Net.Http;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.SignalR.Common.RestClients
+{
+    internal static class HttpExceptionWrapperExtension
+    {
+        internal static Exception WrapAsAzureSingalRException(this Exception e, Uri baseUri)
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:13]: `Singal` --> `Signal`

File: src/Microsoft.Azure.SignalR.Management/ServiceManager.cs [hunk 19]
```diff
@@ -156,31 +154,27 @@ public string GenerateClientAccessToken(string hubName, string userId = null, IL
...
+                return false;
+            }
+            catch (Exception ex)
+            {
+                throw ex.WrapAsAzureSingalRException(_restClient.BaseUri);
+            }
         }
     }
-}
+}
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManager.cs:180]: revert

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs (hunk 12 above)
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:45]: Is it possible that the production info changes? can we input it in ctor of `RestClientBuilder`?

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:45]: If RestClientBuilder is a local field, then each change to connection string should trigger a construction of RestClientBuilder, which makes things more complicated.

File: src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs (hunk 13 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:20]: server will provide more details as similar to `StatusCode(status, new { detail = "This is unhealthy" }` later, just want to confirm if in that case `detail` would work

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/RestClients/HttpExceptionWrapperExtension.cs:20]: `var request = operationException.Request; `  The `request` here is a copy of the origin request, so the `detail` is the response content.  In the case you mentioned it works.

File: src/Microsoft.Azure.SignalR.Common/RestClients/RestClientBuilder.cs [hunk 20]
```diff
@@ -12,22 +12,26 @@ internal class RestClientBuilder
...
+            {
+                throw new ArgumentException($"'{nameof(userAgent)}' cannot be null or whitespace", nameof(userAgent));
+            }
+
+            var azUserAgentHandler = new AsrsUserAgentHandler(userAgent);
             _handlers = new DelegatingHandler[] { azUserAgentHandler };
-            return this;
         }
 
         internal RestClientBuilder WithRootHandler(HttpClientHandler rootHandler)
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Common/RestClients/RestClientBuilder.cs:37]: public

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Common/RestClients/RestClientBuilder.cs:38]: This method is supposed to be internal when the class is public in the future.

**Review by vicancy — APPROVED** [review]

**Pull request merged**